interface/mmal/vc/mmal_vc_client.c

```diff
@@ -477,30 +676,36 @@ static VCHIQ_STATUS_T mmal_vc_vchiq_callback(VCHIQ_REASON_T reason,
          if (msg_hdr->msgid == MMAL_WORKER_BUFFER_TO_HOST)
          {
             mmal_worker_buffer_from_host *msg = (mmal_worker_buffer_from_host *)msg_hdr;
-            LOG_TRACE("bulk rx aborted: %p, %d", msg->buffer_header.data, msg->buffer_header.length);
-            vcos_assert(msg->drvbuf.client_context->magic == MMAL_MAGIC);
+            MMAL_VC_CLIENT_BUFFER_CONTEXT_T *client_context = mmal_vc_lookup_client_context(msg->drvbuf.client_context);
+            LOG_TRACE("bulk rx aborted: %08x, %d", msg->buffer_header.data, msg->buffer_header.length);
+            vcos_assert(client_context->magic == MMAL_MAGIC);
             msg->buffer_header.flags |= MMAL_BUFFER_HEADER_FLAG_TRANSMISSION_FAILED;
-            msg->drvbuf.client_context->callback(msg);
+            client_context->callback(msg);
          }
          else
          {
             mmal_worker_event_to_host *msg = (mmal_worker_event_to_host *)msg_hdr;
-            MMAL_PORT_T *port = mmal_vc_port_by_number(msg->client_component, msg->port_type, msg->port_num);
+            MMAL_COMPONENT_T *component = lookup_client_component(msg->client_component);
+            MMAL_VC_CLIENT_BUFFER_CONTEXT_T *client_context =
+                  mmal_vc_lookup_client_context(mmal_buffer_header_driver_data(msg->delayed_buffer)->client_context);
+            MMAL_PORT_T *port = mmal_vc_port_by_number(component, msg->port_type, msg->port_num);
 
             vcos_assert(port);
             LOG_DEBUG("event bulk rx aborted");
             msg->delayed_buffer->flags |= MMAL_BUFFER_HEADER_FLAG_TRANSMISSION_FAILED;
-            mmal_buffer_header_driver_data(msg->delayed_buffer)->
-               client_context->callback_event(port, msg->delayed_buffer);
+
+            client_context->callback_event(port, msg->delayed_buffer);
          }
          vchiq_release_message(service, header);
       }
       break;
    case VCHIQ_BULK_TRANSMIT_ABORTED:
       {
          mmal_worker_buffer_from_host *msg = (mmal_worker_buffer_from_host *)context;
-         LOG_INFO("bulk tx aborted: %p, %d", msg->buffer_header.data, msg->buffer_header.length);
-         vcos_assert(msg->drvbuf.client_context->magic == MMAL_MAGIC);
+         MMAL_VC_CLIENT_BUFFER_CONTEXT_T *client_context =
```

Comment:
@6by9 

With `-Werror` and `-Wno-unused-but-set-variable` (the default), compile fails for me (on aarch64 Gentoo, `gcc` 9.2.0) on this line, with:

```
/var/tmp/portage/media-libs/raspberrypi-userland-1.20191025/work/raspberrypi-userland-1.20191025/interface/mmal/vc/mmal_vc_client.c:705:43: error: unused variable ‘client_context’ [-Werror=unused-variable]
  705 |          MMAL_VC_CLIENT_BUFFER_CONTEXT_T *client_context =
      |                                           ^~~~~~~~~~~~~~
```

Adding something similar to

```c
client_context = NULL;
```

at the end of the function suppresses this.

Reply:
Thanks. I was building debug builds and therefore the assert used the variable.
The other approach would be to do the call within the assert. I'll revisit it when I have a chance. 

I've been trying to tidy up the firmware side of it first as it has all the counterpart casts to convert between the IPC and active formats (even though they are actually identical in a 32bit world).